### PR TITLE
Query improvements

### PR DESCRIPTION
**Pull request by ifesdjeen**

### CODE REVIEW

**Comment by ifesdjeen** [comment]
I'll rebase on top of your commits whenever they're merged.

File: src/io/cyanite/query/ast.clj [hunk 1]
```diff
@@ -102,6 +110,23 @@
      nil
      (resolve! series 1 1))))
 
+(defrecord DerivativeOperation [series]
```
Comment (pyr) [inline src/io/cyanite/query/ast.clj]: If you want to do away with mutability, a cool trick here would be:

```clojure
(transform! [this]
  (traverse! "derivative($0)" (fn [[a b]] (- b a)) nil (partition 2 1 (resolve! series 1 1)))
```

**Comment by pyr** [comment]
Looking great!

File: src/io/cyanite/query/ast.clj (hunk 1 above)
Comment (ifesdjeen) [inline src/io/cyanite/query/ast.clj]: Oh I hats exactly what I wanted/needed

**Comment by ifesdjeen** [comment]
Rebased and pushed all the updates / fixes.

File: src/io/cyanite/engine/writer.clj [hunk 2]
```diff
@@ -13,14 +13,19 @@
...
-    this)
+    (let [task (with-schedule [pool 10]
+                 (set-thread-name! "cyanite-snapshot")
+                 (info "starting snapshot.")
+                 (engine/snapshot! this))]
+      (assoc this :task task)))
   (stop [this]
-    this)
+    (when-let [task (:task this)]
+      (.cancel task))
```
Comment (pyr) [inline src/io/cyanite/engine/writer.clj]: I opted out of doing this because on shutdown, the pool component just cancels all tasks.

**Comment by pyr** [comment]
Plenty of small bugs in my previous code, sorry. :+1: on the PR, I'll try to help fixing test today before merging

**Comment by ifesdjeen** [comment]
I've fixed the test locally, gotta push in a sec :) I've made a stupid mistake in last commit

File: src/io/cyanite/engine/writer.clj (hunk 2 above)
Comment (ifesdjeen) [inline src/io/cyanite/engine/writer.clj]: true, haven't thought of that

**Comment by ifesdjeen** [comment]
Ok, should be good to go now :)

**Comment by pyr** [comment]
thanks!

**Pull request merged**